### ISSUE TITLE
[FEATURE REQUEST] Support MFA authentication

### ISSUE THREAD

**Issue by serge-rider** [microsoft/mssql-jdbc#1053]
## Is your feature request related to a problem? If so, please give a short summary of the problem and how the feature would resolve it
Latest driver version supports all autentication modes (btw thanks for the MSI support!) but Active Directory MFA.
Do you have this feature in plans?

We are using SQL Server/Azure driver in our open-source product DBeaver and already have quite many requestss for MFA support which we can't implement because driver itself doesn't support it. 
For example: https://github.com/dbeaver/dbeaver/issues/5754#issuecomment-490537611

**Comment by David-Engel** [comment]
@serge-rider Thanks for filing this issue. MFA is in our backlog but the main challenge in adding it directly to the driver is how to implement the UI for it. Should the driver implement a UI itself? Should it implement some kind of hooks for a UI? If it implements a UI, which Java UI framework(s) should we choose without bloating driver dependencies?

Technically, DBeaver can implement AD MFA in top the JDBC driver using the AccessToken authentication method. DBeaver can obtain the token itself via Azure AD REST APIs and pass the token to the driver. I do realize this comes with the overhead of maintaining that code. But this is exactly how Azure Data Studio has done it themselves on top of System.Data.SqlClient in .NET Core, which also does not support AD MFA itself but does support AccessToken.

**Comment by cheenamalhotra** [comment]
Alos I realized the issue mentioned above (dbeaver/dbeaver#5754 (comment)) is for MSI Authentication, which the driver supports, more info here: https://docs.microsoft.com/en-us/sql/connect/jdbc/connecting-using-azure-active-directory-authentication?view=sql-server-2017#connecting-using-activedirectorymsi-authentication-mode

You must note that MSI only works on client machines that are Azure Resources (Azure VM, Azure Function App, App Service) with enabled "Identity" option. It has nothing to do with MFA. Identity is a separate entity and must be registered as user (from External provider) on Database one intends to connect to.

**Comment by David-Engel** [comment]
The original issue was for MSI, but the issue filer for some reason then pivoted to MFA in the comment which was linked to.

**Comment by cheenamalhotra** [comment]
Yes, that's why I mentioned, even if users can connect with MFA in SSMS doesn't mean they will be able to connect with MSI. They're not related.

You need account credentials for MFA to work whereas Identity is a separate entity and needs to be added to database one is trying to connect to.

**Comment by dam1960** [comment]
Thanks everyone for the insight to this issue.  Here's a link to a doc explaining Azure AD and Authentication with the JDBC driver.  It does not mention MFA as an option.

[](https://docs.microsoft.com/en-us/sql/connect/jdbc/connecting-using-azure-active-directory-authentication?view=sql-server-2017)

**Comment by WarrenFerrell** [comment]
Hoping this gets some more attention as more and more companies are requiring MFA for all access to their data. We are switching from Postgres to AzureSql for various reason (among them MFA) and many of our developers have grown used to Dbeaver. It would make the transition from Postgres much more seamless if they could continue using the same tooling.

**Comment by dam1960** [comment]
@David-Engel Any movement on this?  It is becoming a real problem.  Azure Data Studio is a poor replacement for DBeaver.

**Comment by MikielAgutu** [comment]
+1

**Comment by bdleavitt** [comment]
Any update on this feature? 

Would be great if this could behave similarly to the SQL ODBC Driver with support for AAD interactive. https://docs.microsoft.com/en-us/sql/connect/odbc/using-azure-active-directory?view=sql-server-ver15#new-andor-modified-dsn-and-connection-string-keywords

**Comment by ulvii** [comment]
Hi everyone,
To be able to provide MFA support, we would first need to switch from [adal4j to msal4j](https://github.com/microsoft/mssql-jdbc/issues/1227). The feature is already on our roadmap. There is no timeline yet, will update this issue as soon as we have updates.

**Comment by mocleiri** [comment]
Hello, 

At work we have a spring boot application and found out how to configure it to use the access token to login as a service principal (app registration).

We also found a way to use the device code flow and the Azure SQL Database user_impersonation permission to login MFA users using the access token approach.

I just created this repo: 
https://github.com/mocleiri/device-code-sql-server-driver 

that externalizes the token acquisition logic for 2 cases:

1. login as service principal (where you know the clientId and clientSecret)
2. login as the developer with MFA using device code flow.

It uses msal4j to acquire the access tokens in both cases.  

For the device code flow it creates an html page with the https://microsoft.com/devicelogin url and secret code and then opens chrome on that page.

I built this project so we could connect to the database from within Intellij as developers with MFA which works.

In our actual application we have adal4j and msal4j side by side so I think it would be possible to lift out the two flows I've implemented into the driver before a broader switch between libraries has been completed.

Perhaps of interest to others is that we implement token acquisition in a way where it will re acquire the token if it expires.
